src/_pytest/pytester.py

```diff
@@ -508,7 +512,7 @@ def finalize(self):
         self._sys_modules_snapshot.restore()
```

Comment:
While at it, would you like to replace those "snapshot" implementations by your `monkeypatch` instance?

If not, please go ahead and merge it as this is already an improvement 

(every time I looked at this code I had the same idea to use monkeypatch for that, but kept forgetting to actually do it 🙃)

Reply:
Not really now - basically got side-tracked by this already for #5032.

Reply:
btw: to me it looked just like code that could be improved, assuming that `os.environ` would get restored, but that's actually not the case..  therefore targetting master.